### ISSUE TITLE
Parametrized queries in PostgreSQL

### ISSUE THREAD

**Issue by Mikilll94** [dotnet/efcore#14356]
Hey, I have a question about Entity Framework Core in PostgreSQL. How EFCore creates a parametrized queries for PostgreSQL? For example in Sql Server parametrized queries are created using stored procedures, like this:

exec sp_executesql(query, parameters)

But how is this done in PostgreSQL? Are parametrized queries also created via stored procedures?

EF Core version: 2.1
Database Provider: Npgsql.EntityFrameworkCore.PostgresSQL
Operating system: Windows
IDE: Visual Studio 2017 15.8

**Comment by ajcvickers** [comment]
@roji Thoughts?

**Comment by Mikilll94** [comment]
Ok. So I am clarifying...

I want to know how parametrized queries are implemented in Entity Framework Core with PostgreSQL provider.

In Entity Framework Core with Sql Server provider they are implemented using stored procedures. You can verify that using Sql Server profiler.

But how parametrized queries are implemented Entity Framework Core with PostgreSQL provider? I was looking over the Internet to find the answer but I did not find any information. If somebody knows I will be grateful for help.

**Comment by roji** [comment]
@Mikilll94 I'm still not sure I understand your question, but here's an attempt at an answer. In general, EF Core doesn't exactly provide any specific means for invoking stored procedures - neither in SQL Server nor in PostgreSQL.

To define a stored procedure, you need to drop down to raw SQL and do it in the way that's specific to your database (e.g. `CREATE PROCEDURE).

To then execute a stored procedure, you can use `FromSql()` as detailed [in the EF Core docs](https://docs.microsoft.com/en-us/ef/core/querying/raw-sql). As the examples in those docs show, the SQL executed in `FromSql()` can be parameterized. However, this mechanism isn't specific to  stored procedures in any way - you can use `FromSql()` to invoke a stored procedure (e.g. `CALL my_proc()` in PostgreSQL) just like you can use it to run any normal SQL (e.g. `SELECT * FROM blogs`).

To summarize, `FromSql()` can be used to execute raw SQL, which may or may not be parameterized, and which may or may not be used to invoke a stored procedure. [The official EF Core docs](https://docs.microsoft.com/en-us/ef/core/querying/raw-sql) as well as [this tutorial](http://www.entityframeworktutorial.net/efcore/working-with-stored-procedure-in-ef-core.aspx) will probably explain this further.

**Comment by smitpatel** [comment]
I believe @Mikilll94 is asking how the SQL queries with parameters are sent to server in Postgre.
When the linq query is like,
`db.Blogs.Where(b => b.Id = blogId).ToList()`
For SqlServer provider, when above query is run, this is what SQL Profiler shows
```
exec sp_executesql N'SELECT [b].[Id]
FROM [Blogs] AS [b]
WHERE [b].[Id] = @__blogId_0',N'@__blogId_0 int',@__blogId_0=1
```
So what is mechanism in Postgre which sends parameters. It could be possible that underlying command itself contains parameters and does not require specific procedure like sp_executesql which is on SQL Server only.

**Comment by roji** [comment]
PostgreSQL accepts parameterization in SQL statements without any prefix or function similar to `sp_executesql`. Thus, when a LINQ query such as `db.Blogs.Where(b => b.Id = blogId).ToList()` is executed, the PostgreSQL logs will show:

```sql
SELECT b."Id" FROM "Blogs" AS b WHERE b."Id" = $1
```

$1 is the PostgreSQL placeholder format, which is positional rather than named. Note that at the ADO.NET layer, named placeholders such as `@__blogId_0` are translated to $1, but that's not the user's concern in any way.

**Comment by ajcvickers** [comment]
Thanks @roji. Marking this as answered.
